### PR TITLE
Fix #31303: Make reminder notification behaviour selectable..

### LINKED ISSUE #31303: [Bug]: Irritating behavior: Reminders get send to all users with write access within a shared group.
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [ ] This issue is **not** already reported on Github _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

Firstly, this is a duplicate of https://github.com/nextcloud/calendar/issues/3946. But it fits in this Repo better, since I think the bug lies in the Reminder background job. (Which is in this repo.)

Event reminders are not only sent to the attendees of an event, but to all members of all shared groups which have write access too. This behavior is very confusing for end users.

### Steps to reproduce

1. Go to the user settings. (`index.php/settings/users`)
2. Create a group with the following members:
`peter@example.com`
`max@example.com`
3. Go to the calendar app. (`index.php/apps/calendar/timeGridWeek/now `)
4. Create new calendar
5. Share the newly created calendar with the new group (with write-access!)
6. Create a new event in the calendar.
7. Activate a reminder: 5 mins before the event.
8. Under attendees -> Invite the following people:
`john@example.com`
`jane@example.com`not

### Expected behavior
`john@example.com`, `jane@example.com` and the calendar owner should get a reminder notification.
`peter@example.com` and `max@example.com` should only be able to edit the event and not get a notification, since they're not attendees.

So, only the _attendees_ plus the _event owner_ should get a reminder notification.
But I think that some people also want that group members get a notification too, so this behavior should be configurable in the settings.
And I don't think that `write-access` should have something to do with who gets reminder notifications and who don't.

### Actual behaviour

`peter@example.com`, `max@example.com`, `john@example.com`, `jane@example.com` and the calendar owner get notifications.

Also members of groups which don't have write-access don't get reminders!

### Possible solution
I imagine a toggle switch in  `Settings -> Administration -> Groupware`:
"Send reminder notifications to attendees and also shared group members."

### Installation method

Manual installation (Cloned git repo)

### Operating system

Debian/Ubuntu

### PHP engine version

PHP 7.4

### Web server

Apache (supported)

### Database engine version

SQlite

### Is this bug present after an update or on a fresh install?

Fresh Nextcloud Server install

### Are you using the Nextcloud Server Encryption module?

Encryption is Disabled

### What user-backends are you using?

- [X] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

```shell
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "10.0.2.5"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "sqlite3",
        "version": "24.0.0.4",
        "overwrite.cli.url": "http:\/\/10.0.2.5",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "installed": true,
        "debug": true,
        "app_install_overwrite": [
            "richdocuments",
            "richdocumentscode",
            "announcementcenter",
            "deck"
        ],
        "mail_smtpmode": "smtp",
        "mail_smtpauthtype": "LOGIN",
        "mail_sendmailmode": "smtp",
        "mail_smtpauth": 1,
        "mail_smtpsecure": "tls",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "force_language": "en"
    }
}
```

### List of activated Apps

```shell
Enabled:
  - accessibility: 1.10.0
  - calendar: 3.1.0-alpha.1
  - cloud_federation_api: 1.7.0
  - comments: 1.14.0
  - contactsinteraction: 1.5.0
  - dashboard: 7.4.0
  - dav: 1.22.0
  - deck: 1.6.0
  - federatedfilesharing: 1.14.0
  - federation: 1.14.0
  - files: 1.19.0
  - files_sharing: 1.16.0
  - files_trashbin: 1.14.0
  - files_versions: 1.17.0
  - lookup_server_connector: 1.12.0
  - oauth2: 1.12.0
  - provisioning_api: 1.14.0
  - settings: 1.6.0
  - sharebymail: 1.14.0
  - systemtags: 1.14.0
  - theming: 1.15.0
  - twofactor_backupcodes: 1.13.0
  - updatenotification: 1.14.0
  - user_status: 1.4.0
  - weather_status: 1.4.0
  - workflowengine: 2.6.0
Disabled:
  - admin_audit
  - bruteforcesettings
  - encryption
  - files_external
  - testing
  - user_ldap
```

### Additional info
I think the issue lies in this file:
https://github.com/nextcloud/server/blob/6ab9a13ad9341bcdc1f6fc0ff866e430d5ebf956/apps/dav/lib/CalDAV/Reminder/ReminderService.php#L148

```php
$users = $this->getAllUsersWithWriteAccessToCalendar($reminder['calendar_id']);
$user = $this->getUserFromPrincipalURI($reminder['principaluri']);
if ($user) {
$users[] = $user;
}
```

### PR DESCRIPTION
**Pull request by dzatoah**
Fixes https://github.com/nextcloud/server/issues/31303 and https://github.com/nextcloud/calendar/issues/3946.
 
grafik

I'm open for your opinions. ^^
And please check my spelling, I'm not native. :D

### CODE REVIEW

File: apps/dav/lib/CalDAV/Reminder/ReminderService.php [hunk 1]
```diff
@@ -110,8 +111,10 @@ public function __construct(Backend $backend,
 	 *
 	 * @throws NotificationProvider\ProviderNotAvailableException
 	 * @throws NotificationTypeDoesNotExistException
+	 *
+	 * @param IConfig $config
 	 */
-	public function processReminders():void {
+	public function processReminders(IConfig $config):void {
```
Comment (tcitworld) [inline apps/dav/lib/CalDAV/Reminder/ReminderService.php:117]: IConfig should be injected through autoload in the constructor, not passed here.

File: apps/dav/lib/CalDAV/Reminder/ReminderService.php [hunk 2]
```diff
@@ -141,7 +144,12 @@ public function processReminders():void {
 				continue;
 			}
 
-			$users = $this->getAllUsersWithWriteAccessToCalendar($reminder['calendar_id']);
+			if ($config->getAppValue('dav', 'dontSendEventRemindersToSharedGroupMembers', 'no') === 'no') {
+				$users = $this->getAllUsersWithWriteAccessToCalendar($reminder['calendar_id']);
+			} else {
+				$users = array();
```
Comment (tcitworld) [inline apps/dav/lib/CalDAV/Reminder/ReminderService.php:150]:
```suggestion
				$users = [];
```

File: apps/dav/lib/CalDAV/Reminder/ReminderService.php [hunk 3]
```diff
@@ -141,7 +144,12 @@ public function processReminders():void {
 				continue;
 			}
 
-			$users = $this->getAllUsersWithWriteAccessToCalendar($reminder['calendar_id']);
+			if ($config->getAppValue('dav', 'dontSendEventRemindersToSharedGroupMembers', 'no') === 'no') {
```
Comment (tcitworld) [inline apps/dav/lib/CalDAV/Reminder/ReminderService.php:147]: Rename to `sendEventRemindersToSharedGroupMembers` and use `yes`.

**Review by tcitworld — CHANGES_REQUESTED** [review]

**Comment by tcitworld** [comment]
Maybe we can add left-padding to this element to make it clear it belongs to the same group as above, like in the sharing admin settings?

**Comment by D0n1elT** [comment]
> Maybe we can add left-padding to this element to make it clear it belongs to the same group as above, like in the sharing admin settings?

grafik

Requested changes are made.

File: apps/dav/src/views/CalDavSettings.vue [hunk 4]
```diff
@@ -57,7 +57,29 @@
...
 		</p>
-		<p>
+		<p style="padding-left: 28px">
+			<input id="caldavSendEventRemindersToSharedGroupMembers"
+				v-model="sendEventRemindersToSharedGroupMembers"
+				type="checkbox"
+				class="checkbox"
+				:disabled="!sendEventReminders">
+			<label for="caldavSendEventRemindersToSharedGroupMembers">
+				{{ $t('dav', 'Send reminder notifications not to shared group members.') }}
```
Comment (tcitworld) [inline apps/dav/src/views/CalDavSettings.vue:67]: Might not be group shares but direct user-shares too.

File: apps/dav/lib/CalDAV/Reminder/ReminderService.php (hunk 1 above)
Comment (D0n1elT) [reply apps/dav/lib/CalDAV/Reminder/ReminderService.php]: Could you provide docs about the autoloader, please? I didn't grasp the concept of it yet.

File: apps/dav/src/views/CalDavSettings.vue (hunk 4 above)
Comment (D0n1elT) [reply apps/dav/src/views/CalDavSettings.vue:67]: How would you phrase this?

Comment (D0n1elT) [reply apps/dav/src/views/CalDavSettings.vue:67]: `Send notifications to all shared users of an event and not only to the organizer and attendees.` ?

Comment (tcitworld) [reply apps/dav/src/views/CalDavSettings.vue:67]: Something like `Send notifications to calendar sharees as well`.

Comment (D0n1elT) [reply apps/dav/src/views/CalDavSettings.vue:67]: `Send notifications to calendar sharees as well as to the organizer and attendees.` ?

File: apps/dav/src/views/CalDavSettings.vue [hunk 5]
```diff
@@ -57,7 +57,29 @@
...
+			</label>
+			<br>
+			<!-- Can use v-html as:
+				- $t passes the translated string through DOMPurify.sanitize,
+				- replacement strings are not user-controlled. -->
+			<!-- eslint-disable-next-line vue/no-v-html -->
+			<em v-html="sendEventRemindersHelpText" />
```
[the remark this answers is not in the document]
Comment (D0n1elT) [reply apps/dav/src/views/CalDavSettings.vue:74]: Will be removed.

File: apps/dav/src/views/CalDavSettings.vue (hunk 4 above)
Comment (tcitworld) [reply apps/dav/src/views/CalDavSettings.vue:67]: We can move the second part of the sentence above, something like : `reminders are always sent to organizers and attendees`.

Comment (D0n1elT) [reply apps/dav/src/views/CalDavSettings.vue:67]: Like this?

Comment (D0n1elT) [reply apps/dav/src/views/CalDavSettings.vue:67]: Like this?

grafik

**Comment by D0n1elT** [comment]
One of the failing tests:
```html
FAIL apps/dav/src/views/CalDavSettings.spec.js
  ● Console

    console.error
      [Vue warn]: Property or method "sendEventRemindersToSharedGroupMembers" is not defined on the instance but referenced during render. Make sure that this property is reactive, either in the data option, or for class-based components, by initializing the property. See: https://vuejs.org/v2/guide/reactivity.html#Declaring-Reactive-Properties.
      
      found in
      
      ---> <CalDavSettings>
             <Root>

    > 1 | <template>
        | ^
      2 | 	<div class="section">
      3 | 		<h2>{{ $t('dav', 'Calendar server') }}</h2>
      4 | 		<!-- Can use v-html as:
```

I don't get it, do you know what that means? @tcitworld

**Comment by D0n1elT** [comment]
@tcitworld, @miaulalala, @ChristophWurst

I've had some problems with CI but think this PR is ready for another review now.

File: apps/dav/src/views/CalDavSettings.vue [hunk 6]
```diff
@@ -57,7 +57,21 @@
...
+				v-model="sendEventRemindersToSharedGroupMembers"
+				type="checkbox"
+				class="checkbox"
+				:disabled="!sendEventReminders">
+			<label for="caldavSendEventRemindersToSharedGroupMembers">
+				{{ $t('dav', 'Send reminder notifications to calendar sharees as well') }}
+			</label>
+			<br>
+			<em>
+				{{ $t('dav', 'Reminders are always sent to organizers and attendees.' )}}
```
Comment (tcitworld) [inline apps/dav/src/views/CalDavSettings.vue:71]:
```suggestion
				{{ $t('dav', 'Reminders are always sent to organizers and attendees.') }}
```

File: apps/dav/src/views/CalDavSettings.vue [hunk 7]
```diff
@@ -57,7 +57,21 @@
 				{{ $t('dav', 'Notifications are sent via background jobs, so these must occur often enough.') }}
 			</em>
 		</p>
-		<p>
+		<p style="padding-left: 28px">
```
Comment (tcitworld) [inline apps/dav/src/views/CalDavSettings.vue:60]: Please use a class instead defined in a `<style lang="scss" scoped>` block below.

**Comment by D0n1elT** [comment]
Hey @tcitworld,

Changes are done. Can you review again? ^^

File: apps/dav/src/views/__snapshots__/CalDavSettings.spec.js.snap [hunk 8]
```diff
@@ -126,7 +126,35 @@ exports[`CalDavSettings interactions 1`] = `
       </em>
     </p>
      
-    <p>
+    <p
+      style="padding-left: 28px;"
```
Comment (tcitworld) [inline apps/dav/src/views/__snapshots__/CalDavSettings.spec.js.snap:130]: Was this snapshot updated to reflect latest changes?

**Review by tcitworld — APPROVED** [review]
Just needs to test that the failing test is not a flaky one but LGTM overall.

File: apps/dav/src/views/__snapshots__/CalDavSettings.spec.js.snap (hunk 8 above)
Comment (D0n1elT) [reply apps/dav/src/views/__snapshots__/CalDavSettings.spec.js.snap:130]: Oof, I forgot to compile.

Comment (D0n1elT) [reply apps/dav/src/views/__snapshots__/CalDavSettings.spec.js.snap:130]: Actually, running `make` again produces no diff. ^^
So this should be good to go.

Comment (D0n1elT) [reply apps/dav/src/views/__snapshots__/CalDavSettings.spec.js.snap]: Should be done now.

**Comment by D0n1elT** [comment]
> Just needs to test that the failing test is not a flaky one but LGTM overall.

All checks pass now: LGTM? ^^

**Review by ChristophWurst — APPROVED** [review]
LGTM

File: apps/dav/lib/CalDAV/Reminder/ReminderService.php [hunk 9]
```diff
@@ -145,7 +152,12 @@ public function processReminders():void {
 				continue;
 			}
 
-			$users = $this->getAllUsersWithWriteAccessToCalendar($reminder['calendar_id']);
+			if ($this->config->getAppValue('dav', 'sendEventRemindersToSharedGroupMembers', 'yes') === 'no') {
```
Comment (tcitworld) [inline apps/dav/lib/CalDAV/Reminder/ReminderService.php:155]: Of course I forgot to say we should rename `sendEventRemindersToSharedGroupMembers` to `sendEventRemindersToSharedUsers`, as the share can be done with groups as well as direct users. :see_no_evil:

Comment (D0n1elT) [reply apps/dav/lib/CalDAV/Reminder/ReminderService.php:155]: https://github.com/nextcloud/server/pull/31660/

**Pull request merged**